File: 0394_Decode String/DecodeString.cs
public class Solution{
        public string DecodeString(string s)
        {
            var buffer = new StringBuilder() ;
            var countStack = new Stack<int>();
            var resourceStack = new Stack<string>();

            int index = 0;

            while (index < s.Length)
            {
                if (char.IsDigit(s[index]))
                {
                    int count = 0;
                    while (char.IsDigit(s[index]))
                    {
                        count = count * 10 + (s[index++] - '0');
                    }
                    countStack.Push(count);

                }
                else if (s[index] == '[')
                {
                    resourceStack.Push(buffer.ToString());
                    buffer.Clear();
                    index++;
                }
                else if (s[index] == ']')
                {
                    var sb = new StringBuilder(resourceStack.Pop());
                    var count = countStack.Pop();
                    for (int i = 0; i < count; i++)
                    {
                        sb.Append(buffer);
                    }

                    buffer.Clear();
                    buffer.Append(sb);
                    index++;
                }
                else
                {
                    buffer.Append(s[index++]);
                }
            }

            return buffer.ToString();
        }
    }

File: 0394_Decode String/DecodeString.java
class Solution {
    public String decodeString(String s) {
        Stack<Integer> countStack = new Stack<>();
        Stack<StringBuilder> strStack = new Stack<>();
        StringBuilder sb = new StringBuilder();
        int count = 0;
        int index = 0;
        
        while(index < s.length()) {
            char c = s.charAt(index);
            if(Character.isDigit(c)) {
                count *= 10;
                count += (c - '0');
            }else if(c == '[') {
                strStack.push(sb);
                countStack.push(count);
                count = 0;
                sb = new StringBuilder();
            }else if(c == ']') {
                StringBuilder tmpSb = strStack.pop();
                for(int i=countStack.pop() - 1;i>=0;i--) {
                    tmpSb.append(sb);
                }
                sb = tmpSb;
            }else {
                sb.append(c);
            }
            
            index++;
        }
        
        return sb.toString();
    }
}